lexicon/providers/dreamhost.py

```diff
@@ -0,0 +1,181 @@
+    # If possible filter during the query, otherwise filter after response is received.
+    def _list_records(self, rtype=None, name=None, content=None):
+
+        payload = self._get('dns-list_records')
+
+        if payload.get('result', '') != 'success':
+            raise Exception('unable to get records: %s' % payload)
+
+        resource_list = payload.get('data', None)
+        if resource_list is None:
+            raise Exception('unable to get records: %s' % payload)
+
+        resource_list = [
+            resource for resource in resource_list if resource['zone'] == self.domain]
+        if rtype:
+            resource_list = [
+                resource for resource in resource_list if resource['type'] == rtype]
+        if name:
+            name = self._full_name(name)
+            resource_list = [
+                resource for resource in resource_list if resource['record'] == name]
+        if content:
+            resource_list = [
+                resource for resource in resource_list if resource['value'] == content]
+
+        processed_records = []
+        for resource in resource_list:
+            processed_records.append({
+                'id': resource['type'] + '|' + resource['record'] + '|' + resource['value'],
+                'type': resource['type'],
+                'name': resource['record'],
+                'content': resource['value'],
+            })
+
+        return processed_records
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        if identifier == '':
+            identifier = None
+        if identifier is not None:
+            id_list = identifier.split('|')
+            if len(id_list) != 3:
+                raise Exception('invalid identifier: %s' % (identifier))
+
+            id_type, id_name, id_content = id_list
+
+            try:
+                self._delete_record(rtype=id_type, name=id_name, content=id_content)
+            except:  # pylint: disable=bare-except
+                pass
+
+        return self._create_record(rtype=rtype, name=name, content=content)
+
+    # Delete an existing record.
+    # If record does not exist, do nothing.
+    def _delete_record(self, identifier=None, rtype=None, name=None, content=None):
+
+        id_type = None
+        id_name = None
+        id_content = None
+        if identifier is not None:
+            id_list = identifier.split('|')
+            if len(id_list) != 3:
+                raise Exception('invalid identifier: %s' % (identifier))
+
+            id_type, id_name, id_content = id_list
+
+        if rtype is None:
+            rtype = id_type
+        if name is None:
+            name = id_name
+        if content is None:
+            content = id_content
+
+        if name is not None:
+            name = self._full_name(name)
+
+        to_deletes = []
+        if rtype is None or name is None or content is None:
+            if identifier is None:
```

Comment:
I think we can improve in general the clarity of the branching in the `delete_record` method, because I think it becomes a little difficult to understand in which state we are at this point. By making a high level if condition on top of ` delete_record` to test the nullity of `identifier`, you would have two clear and separated logic for the two situations: if `identifier` is set, and if it is not. See also with my proposition about using a hash-based ID, that will influence the design of `delete_record`.

Reply:
I'll change to either we delete based on identifier if identifier is set, or we fallback to list the records based on rtype, name, content and delete the records.